File: server/src/routes/creditCard.ts
import { Router, Request, Response } from "express";
import prisma from "../db";

const router = Router();

router.get("/statements", async (_req: Request, res: Response) => {
  const statements = await prisma.creditCardStatement.findMany({
    include: { account: true },
    orderBy: { statementDate: "desc" },
  });
  res.json(statements);
});

router.get("/statements/:id/transactions", async (req: Request, res: Response) => {
  const id = req.params.id as string;
  const transactions = await prisma.transaction.findMany({
    where: { statementId: id },
    include: { category: true },
    orderBy: { date: "desc" },
  });
  res.json(transactions);
});

router.get("/summary", async (req: Request, res: Response) => {
  const { month, year } = req.query;

  const where: any = {
    account: { type: "credit_card" },
  };

  if (month && year) {
    const startDate = new Date(parseInt(year as string), parseInt(month as string) - 1, 1);
    const endDate = new Date(parseInt(year as string), parseInt(month as string), 0, 23, 59, 59);
    where.date = { gte: startDate, lte: endDate };
  }

  const transactions = await prisma.transaction.findMany({
    where,
    include: { category: true },
  });

  // Group by category
  const categoryMap: Record<string, { name: string; color: string; total: number }> = {};
  let totalSpend = 0;

  for (const t of transactions) {
    if (t.type === "credit") continue; // skip payments/refunds
    if (t.category?.type === "transfer") continue; // skip CC payments

    const catName = t.category?.name || "Uncategorized";
    const catColor = t.category?.color || "#6B7280";

    if (!categoryMap[catName]) {
      categoryMap[catName] = { name: catName, color: catColor, total: 0 };
    }
    categoryMap[catName].total += t.amount;
    totalSpend += t.amount;
  }

  res.json({
    totalSpend,
    byCategory: Object.values(categoryMap).sort((a, b) => b.total - a.total),
  });
});

export default router;

File: server-py/app/routers/credit_card.py
from fastapi import APIRouter
from sqlmodel import Session, select

from ..database import engine
from ..models import CreditCardStatement, Transaction, Category, Account
from ..serializers import tx_dict, account_dict, _z
from ..util import date_bounds

router = APIRouter()


def _statement_dict(st: CreditCardStatement, account: Account | None) -> dict:
    return {
        "id": st.id, "accountId": st.accountId, "statementDate": _z(st.statementDate),
        "billingPeriodStart": _z(st.billingPeriodStart), "billingPeriodEnd": _z(st.billingPeriodEnd),
        "totalDue": st.totalDue, "minimumDue": st.minimumDue, "dueDate": _z(st.dueDate),
        "rewardPoints": st.rewardPoints, "account": account_dict(account),
    }


@router.get("/statements")
def list_statements():
    with Session(engine) as s:
        accts = {a.id: a for a in s.exec(select(Account)).all()}
        stmts = s.exec(
            select(CreditCardStatement).order_by(CreditCardStatement.statementDate.desc())
        ).all()
        return [_statement_dict(st, accts.get(st.accountId)) for st in stmts]


@router.get("/statements/{statement_id}/transactions")
def statement_transactions(statement_id: str):
    with Session(engine) as s:
        cats = {c.id: c for c in s.exec(select(Category)).all()}
        txns = s.exec(
            select(Transaction).where(Transaction.statementId == statement_id)
            .order_by(Transaction.date.desc())
        ).all()
        return [tx_dict(t, category=cats.get(t.categoryId)) for t in txns]


@router.get("/summary")
def cc_summary(month: int | None = None, year: int | None = None):
    gte, lt = date_bounds(month, year)
    with Session(engine) as s:
        cats = {c.id: c for c in s.exec(select(Category)).all()}
        cc_ids = [a.id for a in s.exec(
            select(Account).where(Account.type == "credit_card")).all()]
        q = select(Transaction).where(Transaction.accountId.in_(cc_ids))
        if gte:
            q = q.where(Transaction.date >= gte, Transaction.date < lt)
        txns = s.exec(q).all()

    by_cat: dict[str, dict] = {}
    total = 0.0
    for t in txns:
        if t.type == "credit":
            continue
        c = cats.get(t.categoryId) if t.categoryId else None
        if c and c.type == "transfer":
            continue
        name = c.name if c else "Uncategorized"
        color = c.color if c else "#6B7280"
        entry = by_cat.setdefault(name, {"name": name, "color": color, "total": 0.0})
        entry["total"] += t.amount
        total += t.amount
    return {"totalSpend": total, "byCategory": sorted(by_cat.values(), key=lambda x: x["total"], reverse=True)}
